clients/web/src/collections/Collection.js

```diff
@@ -87,46 +129,138 @@ var Collection = Backbone.Collection.extend({
+                this.pageOffsetStack = [];
+            }
             this.offset = 0;
         } else {
             this.params = params || {};
         }
 
+        if (this.filterFunc && !this.append) {
+            this.pageOffsetStack.push(this.offset);
+        }
+
         var limit = this.pageLimit > 0 ? this.pageLimit + 1 : 0;
 
-        var xhr = restRequest({
-            path: this.altUrl || this.resourceName,
-            data: _.extend({
-                limit: limit,
-                offset: this.offset,
-                sort: this.sortField,
-                sortdir: this.sortDir
-            }, this.params)
-        }).done(_.bind(function (list) {
-            if (this.pageLimit > 0 && list.length > this.pageLimit) {
-                // This means we have more pages to display still. Pop off
-                // the extra that we fetched.
-                list.pop();
-                this._hasMorePages = true;
-            } else {
-                this._hasMorePages = false;
-            }
+        var finalList = []; /* will be built up in pieces */
 
-            this.offset += list.length;
+        function fetchListFragment() {
+            var xhr = restRequest({
+                path: this.altUrl || this.resourceName,
+                data: _.extend({
+                    limit: limit,
+                    offset: this.offset,
+                    sort: this.sortField,
+                    sortdir: this.sortDir
+                }, this.params)
+            });
 
-            if (list.length > 0 || reset) {
-                if (this.append && !reset) {
-                    this.add(list);
+            var result = $.when(xhr).then(_.bind(function (list) {
+                if (!_.isArray(list)) {
+                    list = [list];
+                }
+
+                if (this.pageLimit > 0 && list.length > this.pageLimit) {
+                    // This means we have more pages to display still. Pop off
+                    // the extra that we fetched.
+                    list.pop();
+                    this._hasMorePages = true;
                 } else {
-                    this.reset(list);
+                    this._hasMorePages = false;
                 }
-            }
 
-            this.trigger('g:changed');
-        }, this));
-        xhr.girder = {fetch: true};
-        return xhr;
+                var offsetDelta = list.length;
+
+                /*
+                 * If filtering, decorate the list with their pre-filtered
+                 * indexes.  The index will be needed when adjusting the offset.
+                 */
+                if (this.filterFunc) {
+                    var filter = this.filterFunc;
+                    list = (
+                        list
+                            .map(function (x, index) { return [index, x]; })
+                            .filter(function (tuple) {
+                                return filter(tuple[1]);
+                            })
```

Comment:
I just saw your second comment.

While I could keep the lists separate, I would need to build an index so that I can cross reference items between the two.  I think not having to go through that manually in exchange for some variable name reuse is the better tradeoff, here.